### PR TITLE
Encode doubles as `Double`, do bounds checks, and add support for NaN and Infinity values

### PR DESCRIPTION
**Pull request by drvirgilio**
See discussion in dhall-lang issue [#252](https://github.com/dhall-lang/dhall-lang/issues/252). I'm not sure exactly on where the parser for "[-]Infinity" should go so I kind of guessed.

### CODE REVIEW

**Comment by drvirgilio** [comment]
I merged the changes to the standard so I think this is ready to be merged.

File: dhall/src/Dhall/Pretty/Internal.hs [hunk 1]
```diff
@@ -306,8 +305,8 @@ prettyNumber = literal . Pretty.pretty
 prettyNatural :: Natural -> Doc Ann
 prettyNatural = literal . Pretty.pretty
 
-prettyScientific :: Scientific -> Doc Ann
-prettyScientific = literal . Pretty.pretty . show
+prettyDouble :: Double -> Doc Ann
+prettyDouble = literal . Pretty.pretty . show
```
Comment (PierreR) [inline dhall/src/Dhall/Pretty/Internal.hs]: @Gabriel439 I believe `show` is not needed here.

**Comment by Gabriel439** [comment]
@f-f: I'll go ahead and merge this now and add tests later, because I need to sync this repository with the `dhall-lang` test suite first.  The other reason why is that I would like to add the tests to `dhall-lang` first and then mirror them to this repository rather than the other way around

**Comment by Gabriel439** [comment]
Actually, there are test failures on this branch that are being hidden by a CI misconfiguration on my part.  I'm going to fix that first in a separate pull request and then fix the tests on this branch

**Comment by Gabriel439** [comment]
I have a pull request up against this branch with fixes to the test suite: https://github.com/drvirgilio/dhall-haskell/pull/1

**Comment by drvirgilio** [comment]
Finally got around to making a few changes:
1.  Added tests which check a few boundary conditions such as the point where `Integer/toDouble` rounds up to `Infinity`.
2.  Found a workaround for `fromInteger`'s weird rounding behavior by using `read . show`.
3.  Made the changes suggested by @Gabriel439

**Comment by Gabriel439** [comment]
@drvirgilio: Thanks for doing this! 🙂
